disco_aws_automation/disco_elasticsearch_archive.py

```diff
@@ -0,0 +1,514 @@
+        """
+        total_bytes = cluster_stats["nodes"]["fs"]["total_in_bytes"]
+        free_bytes = cluster_stats["nodes"]["fs"]['free_in_bytes']
+        need_bytes = total_bytes * (1 - threshold)
+        logging.debug("disk utilization : %i%%", (total_bytes - free_bytes) * 100 / total_bytes)
+        return need_bytes - free_bytes
+
+    def _shards_to_delete(self, cluster_stats, max_shards):
+        """
+        Return number of shards to delete in order to keep the number of shards in the
+        cluster under max_shards.
+        """
+        return cluster_stats['indices']['shards']['total'] - max_shards
+
+    def _indices_to_delete(self, threshold, max_shards):
+        """
+        Return list of indices that can be deleted so that used disk space would get
+        below the threshold and the number of shards would be under max_shards.
+        Oldest first.
+        """
+        cluster_stats = self.es_client.cluster.stats()
+
+        bytes_to_free = self._bytes_to_free(cluster_stats, threshold)
+        shards_to_delete = self._shards_to_delete(cluster_stats, max_shards)
+
+        logging.debug("Need to free %i bytes.", bytes_to_free)
+        logging.debug("Need to delete %i shards.", shards_to_delete)
+        if bytes_to_free <= 0 and shards_to_delete <= 0:
+            return []
+
+        index_stats = self._get_all_indices_stats()
+        indices_to_delete = []
+        freed_size = 0
+        deleted_shards = 0
+        for index_stat in index_stats:
+            if freed_size >= bytes_to_free and deleted_shards >= shards_to_delete:
+                break
+
+            if self.snapshot_state(index_stat['index']) == 'SUCCESS':
+                indices_to_delete.append(index_stat['index'])
+                freed_size += index_stat['size']
+                deleted_shards += index_stat['shards']
+
+        return indices_to_delete
+
+    def _archivable_indices(self):
+        """ Return list of indices that are older than yesterday's date """
+        # We don't want to return today's index because it might not be complete yet.
+        # We don't want to return yesterday's index because that ensures we don't run into
+        # the time zone issue dealing with finding today's date.
+        yesterday = datetime.date.today() - datetime.timedelta(days=1)
+        yesterday_str = yesterday.strftime('%Y.%m.%d')
+
+        return [index['index']
+                for index in self._get_all_indices_stats()
+                if index['index'][-10:] < yesterday_str]
+
+    def _all_index_names(self):
+        """ Return list of all indices """
+        return [index['index']
+                for index in self._get_all_indices_stats()]
+
+    def _get_all_indices_stats(self):
+        """
+        Return list of all the indices, their sizes and their numbers of shards
+        sorted by the date from oldest to latest
+        """
+        cat_stats = self.es_client.cat.indices(bytes='b', h='index,store.size,pri,rep')
```

Comment:
Now `CatClient.indices()` is used to obtain index size and shards info.
https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.client.CatClient